SI-8131 fixes residual race condition in runtime reflection

xeno-by opened the pull request:
review @retronym

This is the version of yesterday's pull request with `gitSynchronizedIfNotThreadsafe` optimization disabled. Let's get this in first and experiment with optimizations later.

xeno-by committed:
Revert "synchronizes pendingVolatiles"

This reverts commit 000c18a8fac60065747652368dadcd7850532f3f,
because Symbol.isStable is independent from Type.isVolatile
since fada1ef6b315326ac0329d9e78951cfc95ad0eb0.

xeno-by committed:
SI-8131 fixes residual race condition in runtime reflection

Apparently some completers can call setInfo while they’re not yet done,
which resets the LOCKED flag, and makes anything that uses LOCKED to
track completion unreliable. Unfortunately, that’s exactly the mechanism
that was used by runtime reflection to elide locking for symbols that are
known to be initialized.

This commit fixes the problematic lock elision strategy by introducing
an explicit communication channel between SynchronizedSymbol’s and their
completers. Now instead of trying hard to infer whether it’s already
initialized or not, every symbol gets a volatile field that can be queried
to provide necessary information.

xeno-by committed:
FromJavaClassCompleter is now flag-agnostic

A minor evolution of the notion of completion + also a small performance
optimization.

xeno-by committed:
a note about Symbol.typeSignature

No changes to behavior, just documenting internal workings of reflection
a bit more thoroughly, one step at a time.

xeno-by committed:
removes non-determinism in reflection-sync-potpourri

Depending on the environment in which the test is run, s1 can be either
“String” or “java.lang.String”. This is one of the known non-deterministic
behaviors of our reflection, caused by prefix stripping only working for
packages defined in the root mirror. Until we fix this, I suggest we make
the test more lenient.

retronym commented:
I would prefer to remove any unused `markCompleted` infrastructure in master and reintroduce it if/when we sort out the more fine-grained locking.

xeno-by commented:
@retronym Even without the `gilSynchronizedIfNotThreadsafe`, the infrastructure is used for things like https://github.com/xeno-by/scala/commit/ef9667f70cea1ce7e3a4b8acf24b30e007ddd8d5#diff-6cc67f90cb479803e84e8068d49663fbL619.

xeno-by commented:
Hahaha, it looks like today at night, after removing `beingCompleted` and changing `_initializationStatus` to `_initialized`, I wrote `var _initialized = true` instead of `var _initialized = false`. No wonder that _initialized-based optimizations didn't work, lol :)

retronym commented:
PLS SYNCH

scala-jenkins commented:
(kitty-note-to-self: ignore 32895277)
:cat: Synchronaising! :pray:

retronym commented:
The latest commit still exhibits the race:

```
/code/scala4  qbin/scalac test/files/run/reflection-sync-potpourri.scala && (for i in {1..100}; do parallel -j 50 qbin/scala -J-Xmx1G Test ::: {1..10} || break; printf . ;done)
...............Exception in thread "Reflector-733" scala.reflect.internal.Symbols$CyclicReference: illegal cyclic reference involving class Codec
    at scala.reflect.internal.Symbols$Symbol$$anonfun$info$3.apply(Symbols.scala:1369)
    at scala.reflect.internal.Symbols$Symbol$$anonfun$info$3.apply(Symbols.scala:1367)
    at scala.Function0$class.apply$mcV$sp(Function0.scala:40)
    at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
    at scala.reflect.internal.Symbols$Symbol.lock(Symbols.scala:489)
    at scala.reflect.internal.Symbols$Symbol.info(Symbols.scala:1367)
    at scala.reflect.runtime.SynchronizedSymbols$SynchronizedSymbol$$anon$8.scala$reflect$runtime$SynchronizedSymbols$SynchronizedSymbol$$super$info(SynchronizedSymbols.scala:167)
```

xeno-by commented:
@retronym That's too bad - on my machine the torture test ran for quite some time without crashing. Allright, I'll take a look tomorrow.

xeno-by commented:
@retronym Can't reproduce it on my machine. Could we discuss this over a hangout?

viktorklang commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -1976,12 +1976,12 @@ trait Types
+   * with synchronized, because they are accessed only from isVolatile, which is called only from
+   * Typer.
+   */
+  private var volatileRecursions: Int = 0
```
It's quite confusing to name something "volatile" which is not @volatile.

xeno-by replied:
From what I understand, that's the scalac internals terminology.

viktorklang replied:
to name non-volatile variables "volatileX"?

xeno-by replied:
Apparently yes. I have no knowledge in that area of the compiler. /cc @retronym @adriaanm

viktorklang commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -1976,12 +1976,12 @@ trait Types
-  def pendingVolatiles = _pendingVolatiles
+  /* Syncnote: The `volatile` var and `pendingVolatiles` mutable set need not be protected
+   * with synchronized, because they are accessed only from isVolatile, which is called only from
+   * Typer.
```
So they are only read by and written to, by a single given thread?

xeno-by replied:
Effectively, yes, because the compiler is single-threaded, and this code isn't accessed from outside the compiler.

viktorklang replied:
Why do they need to be volatile if only ever read or written to by a single thread?

xeno-by replied:
They are no longer volatile. Previously (in 2.10.x) they had to be, but not in master. That's why this PR removes synchronization for them: https://github.com/xeno-by/scala/commit/b88198ee2343ad7209d1b33ef3914da0d6be118c.

viktorklang commented on src/reflect/scala/reflect/runtime/SynchronizedSymbols.scala:
```diff
@@ -31,23 +32,103 @@ private[reflect] trait SynchronizedSymbols extends internal.Symbols { self: Symb
+     *  of non-concurrent compiler caches, so we need to serialize certain operations on types in order to make
+     *  sure that things stay deterministic). However, in case of symbols there's hope, because it's only during
+     *  initializaton that symbols are thread-unsafe. After everything's set up, symbols become immutable
+     *  (sans a few deterministic caches that can be populated simultaneously by multiple threads) and therefore thread-safe.
```
How is thread-safety achieved by those? (safe publication etc)

viktorklang commented on src/reflect/scala/reflect/runtime/SynchronizedSymbols.scala:
```diff
@@ -31,23 +32,103 @@ private[reflect] trait SynchronizedSymbols extends internal.Symbols { self: Symb
+     *       that effecively guards `Symbol.initialize`), which means that there can't be update conflicts.
+     *    2) If someone reads a stale value of status, then the worst thing that might happen is that this someone
+     *       is going to spuriously call `initialize`, which is either a gil-protected operation (if the symbol isn't inited yet)
+     *       or a no-op (if the symbol is already inited), and that is fine in both cases.
```
Is this different from lazy-val-double-checked-locking, and if so, how and why?

xeno-by replied:
You mean you would like to know whether this contraption can be replaced by a lazy val?

viktorklang replied:
yeah

xeno-by replied:
Initialization of symbols is more complicated than just calling a method. First, for different symbols it starts and ends in different places in the compiler. Second, there are different initialization states. Some symbols are partially initialized upon creation, some symbols are not.

viktorklang replied:
Ah, ok. Thanks for the explanation!

viktorklang commented on src/reflect/scala/reflect/runtime/SynchronizedSymbols.scala:
```diff
@@ -31,23 +32,103 @@ private[reflect] trait SynchronizedSymbols extends internal.Symbols { self: Symb
+     */
+    @volatile private[this] var _initialized = false
+    @volatile private[this] var _initializationMask = TopLevelPickledFlags
+    override def markFlagsCompleted(mask: Long): this.type = { _initializationMask = _initializationMask & ~mask; this }
```
Is this method intended to be thread-safe?

viktorklang commented on src/reflect/scala/reflect/runtime/SynchronizedSymbols.scala:
```diff
@@ -31,23 +32,103 @@ private[reflect] trait SynchronizedSymbols extends internal.Symbols { self: Symb
+    @volatile private[this] var _initialized = false
+    @volatile private[this] var _initializationMask = TopLevelPickledFlags
+    override def markFlagsCompleted(mask: Long): this.type = { _initializationMask = _initializationMask & ~mask; this }
+    override def markAllCompleted(): this.type = { _initializationMask = 0L; _initialized = true; this }
```
Is this method intended to be thread-safe?

xeno-by replied:
I don't think so. Completion status is only updated under a lock, i.e. in a single-threaded fashion.

viktorklang replied:
A common and highly useful practice is to annotate such method with which lock they are used under.

xeno-by replied:
By "annotate" you mean provide a comment, or there's a more robust way of attaching this kind of metadata to the method?

viktorklang replied:
Something like this: https://github.com/jbossas/jboss-threads/blob/master/src/main/java/org/jboss/threads/OrderedExecutor.java#L48

xeno-by committed:
turns off the gilSynchronizedIfNotInited optimization

According to Jason (https://github.com/scala/scala/pull/3391#issuecomment-32904460),
this one is still causing trouble, so we have to turn it off until
we have time to debug it.

retronym commented:
I'm going to merge this one as is. @xeno-by I'd appreciate if you could followup on Viktor's comments in the next round that tries for more fine grained locks.
